### PR TITLE
Use SelectNext for StudentSelector

### PR DESCRIPTION
**Pull request by acelaya**
> Requires https://github.com/hypothesis/frontend-shared/pull/1330

This PR replaces `Select` with `SelectNext` in the `StudentSelector`.

Before:

image

After:

image

`SelectNext` needs to have a "fixed" size to make sure it does not resize when selected option changes, so I have changed its sizes to be:

* 20rem for extra large devices (this was already like that)
* 16rem for large devices
* 12rem for medium devices
* full (the default) for small and extra small devices, as the components collapse in a column there.

[Grabación de pantalla desde 2023-10-18 11-56-35.webm](https://github.com/hypothesis/lms/assets/2719332/34bbc54c-a3d7-4b4a-b3c3-f293ac6c352a)

### Testing steps

Open any assignment and verify students can be selected as with the regular select.

### CODE REVIEW

File: lms/static/scripts/frontend_apps/components/StudentSelector.tsx [hunk 1]
```diff
@@ -86,29 +77,25 @@ export default function StudentSelector<Student extends StudentOption>({
...
             variant="dark"
           />
-          <Select
-            classes="min-w-[12rem] xl:w-[20rem]"
-            aria-label="Select student"
-            id={selectId}
-            onChange={handleSelectStudent}
+          <label className="sr-only" htmlFor={selectId}>
+            Select student
+          </label>
```
Comment (acelaya) [inline lms/static/scripts/frontend_apps/components/StudentSelector.tsx:82]: This is intended to mimic the previous `aria-label="Select student"`.

**Comment by robertknight** [comment]
When testing with VoiceOver I noticed that the text of the selected student was not read when the select control had focus. A minor issue is that when clicking in the iframe, the listbox is not dismissed, although the focus ring does disappear from the selected item. I think we may have discussed this before, but I'm not sure if we found a solution.

Before:

VoiceOver before

Note the text format is `{current value} {label} {control type} {state}`.

After:

VoiceOver text

**Comment by acelaya** [comment]
> When testing with VoiceOver I noticed that the text of the selected student was not read when the select control had focus.

This is probably because of the [label I added](https://github.com/hypothesis/lms/pull/5764#discussion_r1362071363) to mimic the previous `aria-label`. I guess it should be solved by removing it or making sure its content changes dynamically.

What I don't know is why it announces the first part, "Student 1 of 2". That is not linked with the select as far as I can tell 🤔 

I'll check it just in case.

> A minor issue is that when clicking in the iframe, the listbox is not dismissed, although the focus ring does disappear from the selected item. I think we may have discussed this before, but I'm not sure if we found a solution.

We discussed this for the comment popover, but that one is closed only with `useClickAway`, which is tricky to handle cross-frame.

In the case of `SelectNext` we also have `onFocusAway`, but it does not check if the component loses focus, but instead, if the newly focused element is outside of some container.

I gues s that, if teh focus has moved to a different frame, the events don't trigger as one would expect. I will try to put together a sandbox to test this.

**Comment by acelaya** [comment]
> A minor issue is that when clicking in the iframe, the listbox is not dismissed, although the focus ring does disappear from the selected item.

This PR should address this issue https://github.com/hypothesis/frontend-shared/pull/1326

**Comment by acelaya** [comment]
@robertknight I have found out why it now announces "Student 1 of 2". The reason is that it is a `label` linked with the select via `htmlFor`.

Apparently, the previous `aria-label` on the select itself was "overwriting" the other label when announced by the screen reader, while it is now "combined" with the new sr-only label I added, making it announce both.

My suggestion is that we remove the `htmlFor` reference from the uppermost label (the one with the "Student 1 of 2" text) and set a dynamic content in the new sr-only label. That mimics the previous behavior.

The only drawback is that clicking in the "Student 1 of 2" label will no longer focus the select, but considering how "small" it is, I think it's negligible.

What do you think?

**Comment by robertknight** [comment]
Hi @acelaya, I should have been clearer. The inclusion of "Student 1 of 2" in the new caption is not the problematic change here. In fact I think it is fine to leave that as-is. Removing the `htmlFor` on the label would leave a `<label>` with no associated element, which would be semantically incorrect. The problem here is the fact that the current _value_ of the select ("aunltd learner" in the above screenshot) is no longer read before the label.

**Comment by acelaya** [comment]
> Hi @acelaya, I should have been clearer. The inclusion of "Student 1 of 2" in the new caption is not the problematic change here. In fact I think it is fine to leave that as-is. Removing the `htmlFor` on the label would leave a `<label>` with no associated element, which would be semantically incorrect. The problem here is the fact that the current _value_ of the select ("aunltd learner" in the above screenshot) is no longer read before the label.

Gotcha. Then I think just adding the name of the student in the new label should be enough:

```diff
diff --git a/lms/static/scripts/frontend_apps/components/StudentSelector.tsx b/lms/static/scripts/frontend_apps/components/StudentSelector.tsx
index aa9469c0..47cc4eca 100644
--- a/lms/static/scripts/frontend_apps/components/StudentSelector.tsx
+++ b/lms/static/scripts/frontend_apps/components/StudentSelector.tsx
@@ -78,7 +78,7 @@ export default function StudentSelector<Student extends StudentOption>({
             variant="dark"
           />
           <label className="sr-only" htmlFor={selectId}>
-            Select student
+            {selectedStudent?.displayName} Select student
           </label>
           <SelectNext
             classes="md:!w-[12rem] lg:!w-[16rem] xl:!w-[20rem]"
```

With that it announces mostly the same as before, but prefixed with "Student 1 of 2" or whatever the uppermost label says.

**Comment by acelaya** [comment]
@robertknight I have pushed the fixes mentioned in previous comments and updated to frontend-shared v6.9.0, which includes some extra fixes needed for the LMS

**Comment by robertknight** [comment]
Explicitly adding the control value into the label is something that `SelectNext` users shouldn't need to do themselves, otherwise some consumer will forget. When testing this I wondered why I hadn't encountered this issue with the `SelectNext` control in the Export tab in the client. After some testing I found that the client's Import/Export tabs _do_ have the same issue when testing in Chrome or Firefox. Under Safari, it seems the button content is treated as being either part of the control's label or its current value.

Aside from not wanting to make each use of `SelectNext` need to add the current value into the label, there is also just the issue that there should be a logical distinction between the label of a select control and the current value. The way native controls work is that they have:

- A label, which can be obtained from one of several sources (`label for=`, `aria-label`, `aria-labelledby`) according to a priority order
- A current value
- A control type and states

The screen reader then reads these out in order, eg. as `{current value}, {label}, {control type}, {state}`.

Taking the grade input as an example, the properties can be inspected in the Accessibility panel of the Elements tab:

Spinbox properties

I wondered how other libraries we've been looking at handle this. The React Spectrum picker uses _two_ `aria-labelledby` attributes on the listbox button, which gets concatenated by the browser. This is semantically implementing the same solution as you have here, but built into the control itself. This seems like a bit of a hack:

Spectrum picker

I couldn't easily see how Headless UI's example works because their reference page doesn't have an example with a label.

The example at https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-select-only/ on the other hand uses the combobox pattern. That ARIA role allows for a separation between value and label, and is read out correctly by in all 3 browsers. This allows a native `<label>` to be used similar to what we are doing here:

Combobox example

I had a look at Chrome's implementation of the [`<selectlist>`](https://open-ui.org/components/selectlist/) control they are working on which is planned as a native styleable select in future. After enabling the "Enable experimental web platform features" flag in Chrome and inspecting the demo at https://codepen.io/una/pen/PoXbgVW, I see that in the accessibility tree, `<selectlist>` is exposed using a combobox.

Turning our control into a combobox by adding `role=combobox` on the button appears to resolve the issue and map the control into the expected structure in the accessibility tree. After doing that I noted an issue that most of the examples in the SelectNext demos page are missing labels.

In summary:

- I think we should be using the combobox role for `SelectNext`, because it allows for creating a control with a notion of both a label and a "current value", which can be read as appropriate by the screen reader. It also seems to be where efforts towards a native styleable select are headed.
- The `SelectNext` examples in frontend-shared should have labels

**Comment by acelaya** [comment]
Thanks Rob. Super detailed explanation and investigation.

**Comment by acelaya** [comment]
I have tested this branch with the changes from https://github.com/hypothesis/frontend-shared/pull/1330, and it solves the issues with screen readers.

**Comment by acelaya** [comment]
I have updated to frontend-shared 6.9.1, which includes the refactored SelectNext with combobox role.

This PR is ready for review again.

File: lms/static/scripts/frontend_apps/components/StudentSelector.tsx [hunk 2]
```diff
@@ -86,29 +77,21 @@ export default function StudentSelector<Student extends StudentOption>({
             title="Previous student"
             variant="dark"
           />
-          <Select
-            classes="min-w-[12rem] xl:w-[20rem]"
+          <SelectNext
+            classes="md:!w-[12rem] lg:!w-[16rem] xl:!w-[20rem]"
```
Comment (robertknight) [inline lms/static/scripts/frontend_apps/components/StudentSelector.tsx:81]: The need for `!` here is unfortunate. Can you think of a way we could avoid that in future.

File: lms/static/scripts/ui-playground/components/ToolbarPage.tsx [hunk 3]
```diff
@@ -125,15 +125,23 @@ export default function ToolbarPage() {
                               title="previous student"
                               variant="dark"
                             />
-                            <Select
-                              aria-label="Select student"
-                              classes="xl:w-80"
-                              id="student-selector"
+                            <label
```
Comment (robertknight) [inline lms/static/scripts/ui-playground/components/ToolbarPage.tsx:128]: Now that we actually have this UI implemented, I'm not sure there is value in keeping the toolbar page, which is going to get out of sync with the actual implementation. We could decide to remove these prototypes. We'll always have the code in Git if we need to get them back.

**Review by robertknight — APPROVED** [review]
LGTM. The updated `SelectNext` now works as expected.

File: lms/static/scripts/frontend_apps/components/StudentSelector.tsx [hunk 4]
```diff
@@ -86,29 +77,25 @@ export default function StudentSelector<Student extends StudentOption>({
...
           >
-            <option
-              key={'student-all'}
-              selected={!hasSelectedStudent}
-              value={-1}
-            >
-              All Students
-            </option>
+            <SelectNext.Option value={null}>
+              {() => 'All Students'}
```
[the remark this answers is not in the document]
Comment (acelaya) [reply lms/static/scripts/frontend_apps/components/StudentSelector.tsx]: I see this is an old comment. It in fact can be done now, and I used that approach everywhere here 🙂

File: lms/static/scripts/frontend_apps/components/StudentSelector.tsx (hunk 2 above)
Comment (acelaya) [reply lms/static/scripts/frontend_apps/components/StudentSelector.tsx:81]: I will check this again, because I think this was the result of a couple of tests I did, but prefixed classes should have precedence over the non-prefixed and built-in `w-full` one, so if I'm not wrong, it should work the same if these classes are set without the important modifier.

File: lms/static/scripts/ui-playground/components/ToolbarPage.tsx (hunk 3 above)
Comment (acelaya) [reply lms/static/scripts/ui-playground/components/ToolbarPage.tsx:128]: Sounds reasonable.

File: lms/static/scripts/frontend_apps/components/StudentSelector.tsx (hunk 2 above)
Comment (acelaya) [reply lms/static/scripts/frontend_apps/components/StudentSelector.tsx:81]: Yep, I can confirm the `!` are not needed.
